rules/order/src/Rector/ClassLike/OrderMethodsByVisibilityRector.php

```diff
@@ -0,0 +1,142 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\Order\Rector\ClassLike;
+
+use PhpParser\Node;
+use PhpParser\Node\Stmt\ClassLike;
+use PhpParser\Node\Stmt\ClassMethod;
+use Rector\Core\Rector\AbstractRector;
+use Rector\Core\RectorDefinition\CodeSample;
+use Rector\Core\RectorDefinition\RectorDefinition;
+use Rector\Order\StmtOrder;
+
+/**
+ * @see \Rector\Order\Tests\Rector\ClassLike\OrderMethodsByVisibilityRector\OrderMethodsByVisibilityRectorTest
+ */
+final class OrderMethodsByVisibilityRector extends AbstractRector
+{
+    /**
+     * @var string
+     */
+    private const VISIBILITY = 'visibility';
+
+    /**
+     * @var string
+     */
+    private const POSITION = 'position';
+
+    /**
+     * @var StmtOrder
+     */
+    private $stmtOrder;
+
+    public function __construct(StmtOrder $stmtOrder)
+    {
+        $this->stmtOrder = $stmtOrder;
+    }
+
+    public function getDefinition(): RectorDefinition
+    {
+        return new RectorDefinition('Orders method by visibility', [
+            new CodeSample(
+                <<<'PHP'
+class SomeClass
+{
+    protected function protectedFunctionName();
+    private function privateFunctionName();
+    public function publicFunctionName();
+}
+PHP
+
+                ,
+                <<<'PHP'
+class SomeClass
+{
+    public function publicFunctionName();
+    protected function protectedFunctionName();
+    private function privateFunctionName();
+}
+PHP
+
+            ),
+        ]);
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getNodeTypes(): array
+    {
+        return [ClassLike::class];
+    }
+
+    /**
+     * @param ClassLike $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        if ($node instanceof Node\Stmt\Interface_) {
+            return null;
+        }
+
+        $classMethods = [];
+        $classMethodsByName = [];
+        foreach ($node->stmts as $position => $classStmt) {
+            if (! $classStmt instanceof ClassMethod) {
+                continue;
+            }
+
+            /** @var string $classMethodName */
+            $classMethodName = $this->getName($classStmt);
+            $classMethodsByName[$position] = $classMethodName;
+
+            $classMethods[$classMethodName]['name'] = $classMethodName;
+            $classMethods[$classMethodName][self::VISIBILITY] = $this->getVisibilityOrder($classStmt);
+            $classMethods[$classMethodName]['final'] = $classStmt->isFinal();
+            $classMethods[$classMethodName]['static'] = $classStmt->isStatic();
+            $classMethods[$classMethodName][self::POSITION] = $position;
```

Comment:
It seems like adding different value for one key.

What about an array?

```php
$classMethods[$classMethodName] = [
    'name' => $classMethodName,
    self::VISIBILITY => $this->getVisibilityOrder($classStmt),
    'is_final' => $classStmt->isFinal(),
    'is_static' => $classStmt->isStatic(),
    self::POSITION => $position
];
```

Or event better value object, so next use of this is not random array, but clear method calls with clear return types? E.g. `$values['is_static']` vs `$object->isStatic()`

Reply:
Ah, now I see bellow, why it's done this way. It's ok then :+1: